delta/app/src/main/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesPracticeRoutes.scala

```diff
@@ -0,0 +1,162 @@
+    (get & pathPrefix("practice") & pathPrefix("resources")) {
+      extractCaller { implicit caller =>
+        (resolveProjectRef & pathEndOrSingleSlash) { ref =>
+          authorizeFor(ref, Write).apply {
+            (entity(as[GenerationInput])) { input =>
+              generate(ref, input)
+            }
+          }
+        }
+      }
+    }
+
+  // Call the generate method matching the schema input
+  private def generate(project: ProjectRef, input: GenerationInput)(implicit caller: Caller) =
+    input.schema match {
+      case ExistingSchema(schemaId) =>
+        emit(resourcesPractice.generate(project, schemaId, input.resource).flatMap(_.asJson))
+      case NewSchema(schemaSource)  =>
+        emit(
+          generateSchema(project, schemaSource, caller).flatMap { schema =>
+            resourcesPractice.generate(project, schema, input.resource).flatMap(_.asJson)
+          }
+        )
+    }
+
+}
+
+object ResourcesPracticeRoutes {
+
+  type GenerateSchema = (ProjectRef, Json, Caller) => IO[SchemaRejection, SchemaResource]
+
+  sealed private[routes] trait SchemaInput extends Product
+
+  private[routes] object SchemaInput {
+
+    // Validate the generated resource with an existing schema
+    final case class ExistingSchema(id: IdSegment) extends SchemaInput
+
+    // Validate the generated resource with the new schema bundled in the request
+    final case class NewSchema(json: Json) extends SchemaInput
+
+    implicit val schemaInputDecoder: Decoder[SchemaInput] =
+      Decoder.instance { hc =>
+        val value          = hc.value
+        val existingSchema = value.asString.map { s => ExistingSchema(IdSegment(s)) }
+        val newSchema      = NewSchema(value)
+        Right(existingSchema.getOrElse(newSchema))
+      }
+  }
+
+  private val noSchema = ExistingSchema(IriSegment(schemas.resources))
+
+  final private[routes] case class GenerationInput(schema: SchemaInput = noSchema, resource: NexusSource)
+
+  private[routes] object GenerationInput {
+    @nowarn("cat=unused")
+    implicit def generationInputDecoder(implicit decodingOption: DecodingOption): Decoder[GenerationInput] = {
+      implicit val configuration: Configuration             = Configuration.default.withDefaults
+      implicit val nexusSourceDecoder: Decoder[NexusSource] = NexusSource.nexusSourceDecoder
+      deriveConfiguredDecoder[GenerationInput]
+    }
+  }
+
+  def apply(
+      identities: Identities,
+      aclCheck: AclCheck,
+      schemas: Schemas,
+      resourcesPractice: ResourcesPractice,
+      schemeDirectives: DeltaSchemeDirectives
+  )(implicit
+      baseUri: BaseUri,
+      s: Scheduler,
+      cr: RemoteContextResolution,
+      ordering: JsonKeyOrdering,
+      decodingOption: DecodingOption
+  ): ResourcesPracticeRoutes =
+    new ResourcesPracticeRoutes(
+      identities,
+      aclCheck,
+      (project, source, caller) => schemas.createDryRun(project, source)(caller).toBIO[SchemaRejection],
```

Comment:
Having this as a private method in `ResourcesPracticeRoutes` would allow getting rid of the constructor all together? Not super important

Reply:
The idea is to make unit-testing easier because `Schemas` comes with a lot of dependencies (doobie, SHACL engine, ...)

Reply:
Using functions like this makes general navigation of the code harder, as well as this constructor being unnecessary. It should already be easy to use mock implementations in a unit test

Interface segregation is a good way of handling this sort of situation and will mean you won't need the dependencies you mention

Reply:
I am not sure we want to do a remake between OOP and FP in the comments here.
As long as we have the `apply` method here, I don't think it hurts navigability.
The FP way allows better expressivity and better flexibility than introducing a multitude of interfaces.
On the flexibility side, what happens if in another case, you want to generate a schema but only care about one of the fields or just if it succeeds ? Or something else ?
You would have to introduce new interfaces or have your mocks to know about the result before dropping the info.
(Those are just simple examples, there could be others)

Reply:
Every time I want to see what code actually run when we are passing one of these functions, I have to navigate multiple steps and potentially multiple branches. That is definitely harder than just clicking on show implementations

In this instance, using a function instead of an interface gives us nothing and costs a lot, which is why I'm saying we should change it here

Reply:
Also I don't believe there is a conflict between OOP and FP here, I'm not sure how using an interface makes something not functional?

Reply:
Interesting discussion! In these kinds of situations we've found single-method traits to be quite a good compromise in the past. They essentially act as just named functions, sort of like a type alias but first-class. Navigation is still easy and Scala's de-sugaring makes them simple to stub. 

For example if a dependent expects a `trait MyTrait { def foo(i: Int): String }`, you can pass it as a stub like `(i: Int) -> "hello"` in tests and the compiler will understand it needs to do the whole `new MyTrait { ... }` bit for you, which is nice. In Scala 3 there's even less boilerplate needed (better type inference).

Not sure I understand this bit `you want to generate a schema but only care about one of the fields or just if it succeeds ?` - what would that look like? (Just trying figure out what I might do in that situation)

As an aside, in Haskell there are a number of approaches to this type of dependency injection. You might do a type alias, newtype wrapper round the function, or a type class (if doing tagless final). Otherwise type signatures become very hard to read when containing lots of functions passed as arguments. The single-method trait isn't that different from the type class pattern as I see it. (Well really all of these approaches are pretty similar 😛 )

The "method dictionary" pattern can be quite neat too - you basically have a bunch of newtype wrapped functions stored in some `Env` data type, and access them by the record accessor (just the thing inside the newtype).  For convenience you might also have syntax to pull out the functions and run them in one fell swoop. In this case that may require quite a big refactor though so I'm not sure.

Reply:
The only downside with SAM (the example `(i: Int) -> "hello"` from above) is that Intellij doesn't really understand that it's an implementation of the interface from a code-navigation perspective. I suppose mostly we would only use SAM in tests, where it isn't so necessary to have the navigation functionality

In production code it's just nice to easily see what the dependency actually does

Reply:
I would love to be able to use a SAM here and we should use them where we can.
The Schemas module is complex and contains a collection of operations though (with like 3 operations to create a schema) and the part which is mentioned is only used in this context (so far).
As it is a matter that goes further than the context of this PR, we should rather discuss that when we are all in the office